text/0000-decorator-support.md

```diff
@@ -0,0 +1,142 @@
+Ember and its users, and while it is not ideal, it allows us to continue moving
+forward and innovating as a framework.
+
+## Detailed design
+
+### Classic Classes
+
+The first thing to note is that Ember.js will continue to work with classic
+classes and classic class syntax for the forseeable future, as a way for users
+to opt-out if they don't want to take the risk of using native classes. This
+includes:
+
+- Classic Classes
+- Classic Components
+- Mixins
+- Computed Properties
+- Observers
+- Tracked Properties
+- Injections
+- All existing classes defined with classic class syntax
+
+Notably, `GlimmerComponent` will _not_ support classic class syntax, due to its
+constraint of having to support both Ember.js _and_ Glimmer.js, where the
+classic class model is not available. However, creating an API compatible
+classic class version using Ember's component manager API should be possible if
+users want to write Glimmer-like component's with classic syntax.
+
+### Decorator Versioning and Support
+
+Because the decorators spec has changed while remaining in a stage 2, it's no
+longer enough to refer to "stage 2" as a version. Since Babel will be providing
+the transforms, we will use their versioning system. [This is currently in
+development](https://github.com/babel/babel/pull/9360), but looks like it will
+be based on the TC39 meeting that the spec was presented at. For instance, the
+current transforms available in Babel are referred to as `nov-2018`.
+
+Ember will begin by supporting the _latest_ version of the decorators transform
+provided by Babel. Currently this is the `nov-2018` version, but by the time
+decorators are released this may change.
+
+The decorators version will be configurable on a _per-app/engine/addon_ basis.
+This will allow the ecosystem to adopt decorators without fears that if an
+application updates, it could break their decorators. If no version is
+specified, the transform will default to the latest version supported by Ember
+at that time. In many cases the only decorators in use will be Ember's, so this
+allows the ecosystem to upgrade without additional maintenance costs. The exact
+method for specifying the version will be determined during implementation, but
+will likely be a configuration option or addon.
+
+When new transforms are released, Ember will continue to support older versions
+of the transforms alongside the new ones until _at least_ the next LTS release,
+with deprecation warnings for any users of the older transforms. This will mean
+shipping multiple different versions of decorators at once, and providing a way
+for apps/addons to import the correct version. Additionally, Ember will attempt
+to provide addon authors with tools that allow _them_ to do this as well, so
+they can support multiple versions of transforms easily and transparently. The
+exact API for these tools is out of scope for this RFC, and will likely be
```

Comment:
I don't believe we can make that kind of guarantee at this point, which is why this RFC is defining the semantics of what we want to support and not focusing on the specifics of the implementation. It may be that supporting multiple versions of decorators does require larger asset sizes, and it will be the responsibility of each addon/app author to weigh the pros and cons of adding more library code to their payloads.

We could maybe outline user stories with the general goals of what we want to achieve, with a caveat that this will be a best-effort goal and not a hard requirement.